cardano-db-sync/src/Cardano/DbSync/Cache/FIFO.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Cardano.DbSync.Cache.FIFO (
+  FifoCache (..),
+  TxIdCache,
+  HashableTxId (..),
+  empty,
+  insert,
+  lookup,
+  getSize,
+  getCapacity,
+  emptyTxIdCache,
+  insertTxIdCache,
+  lookupTxIdCache,
+  cleanupTxIdCache,
+) where
+
+import Cardano.Crypto.Hash.Class (hashToBytes)
+import qualified Cardano.Db as DB
+import Cardano.Ledger.Crypto (StandardCrypto)
+import Cardano.Ledger.SafeHash (extractHash)
+import Cardano.Ledger.TxIn (TxId (..))
+import Cardano.Prelude hiding (empty)
+import qualified Data.HashMap.Strict as HM
+import Data.Hashable (Hashable, hashWithSalt)
+import qualified Data.List.NonEmpty as NE
+
+{- FifoCache: A high-performance, fixed-capacity FIFO (First-In-First-Out) cache implementation
+
+Performance characteristics:
+1. Space efficiency:
+   - Uses UNPACK pragma for maxCapacity and currentSize, reducing memory overhead
+   - Compact representation using NonEmpty for the queue and HashMap for the cache
+
+2. Time complexity:
+   - Lookup: O(1) average case, leveraging HashMap's constant-time lookups
+   - Insert: O(1) amortized time
+     - HashMap insert/delete: O(1) average case
+     - NonEmpty cons/uncons: O(1)
+   - When cache is full:
+     - Efficiently removes oldest item using NonEmpty's O(1) uncons operation
+     - Maintains FIFO order without full list traversal
+
+3. Optimizations:
+   - Uses strict fields (!) to prevent space leaks
+   - Leverages NonEmpty for the queue, ensuring type-safe non-empty invariant
+   - Employs HashMap for fast key-value lookups and updates
+
+4. Scalability:
+   - Performs well with increasing cache sizes due to HashMap's efficiency
+   - Constant-time operations regardless of cache occupancy
+
+5. Memory usage:
+   - Bounded by maxCapacity, preventing unbounded growth
+   - UNPACK pragma reduces memory footprint for frequently accessed fields
+
+6. Thread safety:
+   - Note: This implementation is not thread-safe. Concurrent access should be
+     managed externally if required.
+-}
+
+data FifoCache k v = FifoCache
+  { maxCapacity :: {-# UNPACK #-} !Natural
+  , currentSize :: {-# UNPACK #-} !Int
+  , queue :: !(Maybe (NE.NonEmpty k))
```

Comment:
Is there any reason to use Maybe NonEmpty instead of List?

Reply:
What is the reason for queue?

Reply:
Doesn't `uncons` return the last element inserted?

Reply:
no it's the first element in the list which is the oldest in the FIFO queue

Reply:
`NonEmpty` is simply a list which can't be empty. It's not FIFO.

Reply:
I've removed it